### PR TITLE
add avl tree and heap part cpp code

### PR DESCRIPTION
**Pull request by what-is-me**
- the code was tested. 
- most of them were translated from Java version.
- heap.cpp wasn't translated well, but I have no better idea.

### CODE REVIEW

File: codes/cpp/chapter_heap/my_heap.cpp [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    void print() {
+        cout << "堆的数组表示：" << endl;
+        PrintUtil::printVector(this->maxHeap);
+        cout << "堆的树状表示：" << endl;
+        TreeNode *tree = vecToTree(this->maxHeap);
+        PrintUtil::printTree(tree);
+        freeMemoryTree(tree);
+    }
+};
+void testPush(MaxHeap &maxHeap, int val) {
```
Comment (krahets) [inline codes/cpp/chapter_heap/my_heap.cpp:128]: Add an empty line

File: codes/cpp/chapter_tree/avl_tree.cpp [hunk 2]
```diff
@@ -0,0 +1,220 @@
+/**
+ * File: avl_tree.cpp
+ * Created Time: 2023-2-3
+ * Author: what-is-me (whatisme@outlook.jp)
+ */
+
+#include "../include/include.hpp"
+/* AVL 树 */
```
Comment (krahets) [inline codes/cpp/chapter_tree/avl_tree.cpp:8]: Add an empty line

File: codes/cpp/chapter_tree/avl_tree.cpp [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+class AVLTree {
+public:
+    TreeNode* root;  // 根节点
+private:
+    /* 更新结点高度 */
+    void updateHeight(TreeNode* node) {
+        // 结点高度等于最高子树高度 + 1
+        node->height = max(height(node->left), height(node->right)) + 1;
+    }
+    /* 右旋操作 */
```
Comment (krahets) [inline codes/cpp/chapter_tree/avl_tree.cpp:18]: Add an empty line

File: codes/cpp/chapter_tree/avl_tree.cpp [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        // 以 child 为原点，将 node 向左旋转
+        child->left = node;
+        node->right = grandChild;
+        // 更新结点高度
+        updateHeight(node);
+        updateHeight(child);
+        // 返回旋转后子树的根节点
+        return child;
+    }
+    /* 执行旋转操作，使该子树重新恢复平衡 */
```
Comment (krahets) [inline codes/cpp/chapter_tree/avl_tree.cpp:45]: Add an empty line

Comment (krahets) [reply codes/cpp/chapter_tree/avl_tree.cpp:45]: Same as others.

File: codes/cpp/chapter_tree/avl_tree.cpp [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    PrintUtil::printTree(tree.root);
+}
+
+void testRemove(AVLTree& tree, int val) {
+    tree.remove(val);
+    cout << "\n删除结点 " << val << " 后，AVL 树为" << endl;
+    PrintUtil::printTree(tree.root);
+}
+int main() {
+    AVLTree avlTree;
```
Comment (krahets) [inline codes/cpp/chapter_tree/avl_tree.cpp:194]: Add the comment `/* 初始化空 AVL 树 */`

File: codes/cpp/chapter_heap/heap.cpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    maxHeap.testPoll();
+
+    /* 获取堆大小 */
+    int size = maxHeap.size();
+    cout << "\n堆元素数量为 " << size << endl;
+
+    /* 判断堆是否为空 */
+    bool isEmpty = maxHeap.empty();
+    cout << "\n堆是否为空 " << isEmpty << endl;
+    vector<int> tmp({1, 3, 2, 5, 4});
```
Comment (krahets) [inline codes/cpp/chapter_heap/heap.cpp:66]: Add the comments.

```
    /* 输入列表并建堆 */
    // 时间复杂度为 O(n) ，而非 O(nlogn)
```

File: codes/cpp/chapter_heap/my_heap.cpp [hunk 7]
```diff
@@ -0,0 +1,168 @@
+#include "../include/include.hpp"
+/* 最大堆类 */
```
Comment (krahets) [inline codes/cpp/chapter_heap/my_heap.cpp:2]: Add an empty line

File: docs/chapter_heap/heap.md [hunk 8]
```diff
@@ -85,7 +85,49 @@ comments: true
...
+    heap.pop();         //出队
+    peek = maxHeap.top();  // 4
+    heap.pop();         //出队
+    peek = maxHeap.top();  // 3
+    heap.pop();         //出队
+    peek = maxHeap.top();  // 2
+    heap.pop();         //出队
+    peek = maxHeap.top();  // 1
+    heap.pop();         //出队
+
```
Comment (krahets) [inline docs/chapter_heap/heap.md:117]: Remove the duplicated empty line

File: docs/chapter_heap/heap.md [hunk 9]
```diff
@@ -255,7 +297,23 @@ comments: true
 === "C++"
 
     ```cpp title="my_heap.cpp"
+    // 使用vector而非数组，这样无需考虑扩容问题
```
Comment (krahets) [inline docs/chapter_heap/heap.md:300]: 使用 vector 而非数组

Add a space between English char and Chinese char.

**Review by krahets — COMMENTED** [review]
Thanks for the PR! Please address the comments.

**Comment by krahets** [comment]
Could you help paste the console output to this PR?

**Comment by what-is-me** [comment]
### avl_tree.cpp
```
插入结点 1 后，AVL 树为
——— 1

插入结点 2 后，AVL 树为
    /——— 2
——— 1

插入结点 3 后，AVL 树为
    /——— 3
——— 2
    \——— 1

插入结点 4 后，AVL 树为
        /——— 4
    /——— 3
——— 2
    \——— 1

插入结点 5 后，AVL 树为
        /——— 5
    /——— 4
   |    \——— 3
——— 2
    \——— 1

插入结点 8 后，AVL 树为
        /——— 8
    /——— 5
——— 4
   |    /——— 3
    \——— 2
        \——— 1

插入结点 7 后，AVL 树为
        /——— 8
    /——— 7
   |    \——— 5
——— 4
   |    /——— 3
    \——— 2
        \——— 1

插入结点 9 后，AVL 树为
            /——— 9
        /——— 8
    /——— 7
   |    \——— 5
——— 4
   |    /——— 3
    \——— 2
        \——— 1

插入结点 10 后，AVL 树为
            /——— 10
        /——— 9
       |    \——— 8
    /——— 7
   |    \——— 5
——— 4
   |    /——— 3
    \——— 2
        \——— 1

插入结点 6 后，AVL 树为
            /——— 10
        /——— 9
       |    \——— 8
    /——— 7
   |   |    /——— 6
   |    \——— 5
——— 4
   |    /——— 3
    \——— 2
        \——— 1

插入结点 7 后，AVL 树为
            /——— 10
        /——— 9
       |    \——— 8
    /——— 7
   |   |    /——— 6
   |    \——— 5
——— 4
   |    /——— 3
    \——— 2
        \——— 1

删除结点 8 后，AVL 树为
            /——— 10
        /——— 9
    /——— 7
   |   |    /——— 6
   |    \——— 5
——— 4
   |    /——— 3
    \——— 2
        \——— 1

删除结点 5 后，AVL 树为
            /——— 10
        /——— 9
    /——— 7
   |    \——— 6
——— 4
   |    /——— 3
    \——— 2
        \——— 1

删除结点 4 后，AVL 树为
        /——— 10
    /——— 9
   |    \——— 7
——— 6
   |    /——— 3
    \——— 2
        \——— 1

查找到的结点对象为 0x2aae3d73c60，结点值 = 7
```
### heap.cpp
```
以下测试样例为大顶堆

元素 1 入堆后
堆的数组表示：
[1]
堆的树状表示：
——— 1

元素 3 入堆后
堆的数组表示：
[3, 1]
堆的树状表示：
——— 3
    \——— 1

元素 2 入堆后
堆的数组表示：
[3, 1, 2]
堆的树状表示：
    /——— 2
——— 3
    \——— 1

元素 5 入堆后
堆的数组表示：
[5, 3, 2, 1]
堆的树状表示：
    /——— 2
——— 5
    \——— 3
        \——— 1

元素 4 入堆后
堆的数组表示：
[5, 4, 2, 1, 3]
堆的树状表示：
    /——— 2
——— 5
   |    /——— 3
    \——— 4
        \——— 1

堆顶元素为 5

元素 5 出堆后
堆的数组表示：
[4, 3, 2, 1]
堆的树状表示：
    /——— 2
——— 4
    \——— 3
        \——— 1

元素 4 出堆后
堆的数组表示：
[3, 1, 2]
堆的树状表示：
    /——— 2
——— 3
    \——— 1

元素 3 出堆后
堆的数组表示：
[2, 1]
堆的树状表示：
——— 2
    \——— 1

元素 2 出堆后
堆的数组表示：
[1]
堆的树状表示：
——— 1

元素 1 出堆后
堆的数组表示：
[]
堆的树状表示：

堆元素数量为 0

堆是否为空 1

输入列表并建立小顶堆后
堆的数组表示：
[1, 3, 2, 5, 4]
堆的树状表示：
    /——— 2
——— 1
   |    /——— 4
    \——— 3
        \——— 5
```
### myheap.cpp
```

输入列表并建堆后
堆的数组表示：
[9, 8, 6, 6, 7, 5, 2, 1, 4, 3, 6, 2]
堆的树状表示：
        /——— 2
    /——— 6
   |    \——— 5
   |        \——— 2
——— 9
   |        /——— 6
   |    /——— 7
   |   |    \——— 3
    \——— 8
       |    /——— 4
        \——— 6
            \——— 1

堆顶元素为 9

元素 7 入堆后
堆的数组表示：
[9, 8, 6, 6, 7, 5, 2, 1, 4, 3, 6, 2, 7]
堆的树状表示：
        /——— 2
    /——— 6
   |   |    /——— 7
   |    \——— 5
   |        \——— 2
——— 9
   |        /——— 6
   |    /——— 7
   |   |    \——— 3
    \——— 8
       |    /——— 4
        \——— 6
            \——— 1

堆顶元素 9 出堆后
堆的数组表示：
[8, 7, 6, 6, 7, 5, 2, 1, 4, 3, 6, 2]
堆的树状表示：
        /——— 2
    /——— 6
   |    \——— 5
   |        \——— 2
——— 8
   |        /——— 6
   |    /——— 7
   |   |    \——— 3
    \——— 7
       |    /——— 4
        \——— 6
            \——— 1

堆元素数量为 12

堆是否为空 0

```
I will modify my commit soon.

**Review by krahets — APPROVED** [review]
Thanks!

**Comment by krahets** [comment]
Hi! I found some mistakes in the `my_heap.cpp`. Please check the latest code in https://github.com/krahets/hello-algo/pull/321

**Comment by what-is-me** [comment]
Sorry, I just copied all the name of class and function from the Java code.

**Comment by krahets** [comment]
Except for the naming, the code worked incorrectly.

For example, this result is not correct.

```
元素 7 入堆后
堆的数组表示：
[9, 8, 6, 6, 7, 5, 2, 1, 4, 3, 6, 2, 7]
堆的树状表示：
        /——— 2
    /——— 6
   |   |    /——— 7
   |    \——— 5
   |        \——— 2
——— 9
   |        /——— 6
   |    /——— 7
   |   |    \——— 3
    \——— 8
       |    /——— 4
        \——— 6
            \——— 1
```

**Comment by what-is-me** [comment]
The bug caused by the function swap? 
Sorry I didn't paid attention to it.
I thought the swap should match the class method.

**Comment by what-is-me** [comment]
I'm so sorry I made something wrong in avl_tree.md.
I ignored that my code editor wrongly formated the markdown file and now the page went wrong.
I fixed it in my new rp #323

**Pull request merged**